### ISSUE TITLE
Support restart from checkpoint

### ISSUE THREAD

**Issue by mseeger** [awslabs/keys_values#101]
**Is your feature request related to a problem? Please describe.**
At present, when storing a checkpoint, we do not store the optimizer and learning rate scheduler states properly. This means that restarting from a checkpoint is not really supported.

**Describe the solution you'd like**

- Store complete training state along with every checkpoint. Includes states of optimizer, learning rate scheduler, data iterators
- Optimizer states can be large. Only store them for the last recent K checkpoints
- Support resume from checkpoint
- Test: Stop and resume must give same behavior as not stopping at all

**Comment by mseeger** [comment]
Branch: `restart_checkpoint`

**Comment by mseeger** [comment]
## Testing:
- Run on 2 devices (to test multi-device)
- Run 32k setup to write some checkpoints
- Restart from step-000005, step-000015, and compare

### Initial run: Eval and store every 5 steps:

```
CUDA_VISIBLE_DEVICES="0,1" \
PYTORCH_ALLOC_CONF=expandable_segments:True \
KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/debug14/logs" \
python3 keys_values/__main__.py finetune_long_lora \
    Qwen/Qwen3-4B-Instruct-2507 \
    --out_dir /home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/debug14 \
    --precision bf16-true \
    --verbose some \
    --devices 2 \
    --data Helmet \
        --data.dataset_key nq \
        --data.max_length 32k \
        --data.metadata_dir /home/ubuntu/out/finetune/data \
        --data.trainloader_longest_first True \
    --train.save_interval 5 \
        --train.micro_batch_size 2 \
        --train.epochs 5 \
        --train.average_loss_per_batch True \
    --eval.interval 5 \
        --eval.initial_validation False \
        --eval.use_sample_metric False \
    --kv_cache.cache_length 32768 \
        --kv_cache.chunk_size 2048 \
        --kv_cache.name lastrec-torch-quantized8 \
    --grad.layers_per_cell 2 \
        --grad.layercp_qname default \
        --grad.cachecp_qname torch-quantized8 \
    --optimizer.name AdamW \
        --optimizer.learning_rate 0.0005 \
    --training_state_num 5

[1: 764202.pts-0.ip-172-31-26-205: running...]
```

- Does it store train_state properly? OK
- Does it remove train_state? OK
- Copy `step-000005`, `step-000015`, so they do not get removed!

### Resume from step-000005. No further checkpoints are written:

```
CUDA_VISIBLE_DEVICES="2,3" \
PYTORCH_ALLOC_CONF=expandable_segments:True \
KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/debug14/logs/resume5" \
python3 keys_values/__main__.py finetune_long_lora \
    Qwen/Qwen3-4B-Instruct-2507 \
    --out_dir /home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/debug14 \
    --precision bf16-true \
    --verbose some \
    --devices 4 \
    --data Helmet \
        --data.dataset_key nq \
        --data.max_length 32k \
        --data.metadata_dir /home/ubuntu/out/finetune/data \
        --data.trainloader_longest_first True \
    --train.save_interval 1000 \
        --train.micro_batch_size 2 \
        --train.epochs 5 \
        --train.average_loss_per_batch True \
    --eval.interval 5 \
        --eval.initial_validation False \
        --eval.use_sample_metric False \
    --kv_cache.cache_length 32768 \
        --kv_cache.chunk_size 2048 \
        --kv_cache.name lastrec-torch-quantized8 \
    --grad.layers_per_cell 2 \
        --grad.layercp_qname default \
        --grad.cachecp_qname torch-quantized8 \
    --optimizer.name AdamW \
        --optimizer.learning_rate 0.0005 \
    --resume step-000005

[TODO]
```

### Resume from step-000015. No further checkpoints are written:

```
CUDA_VISIBLE_DEVICES="2,3" \
PYTORCH_ALLOC_CONF=expandable_segments:True \
KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/debug14/logs/resume5" \
python3 keys_values/__main__.py finetune_long_lora \
    Qwen/Qwen3-4B-Instruct-2507 \
    --out_dir /home/ubuntu/out/finetune/neurips_exp/lora/qwen3_4b/debug14 \
    --precision bf16-true \
    --verbose some \
    --devices 4 \
    --data Helmet \
        --data.dataset_key nq \
        --data.max_length 32k \
        --data.metadata_dir /home/ubuntu/out/finetune/data \
        --data.trainloader_longest_first True \
    --train.save_interval 1000 \
        --train.micro_batch_size 2 \
        --train.epochs 5 \
        --train.average_loss_per_batch True \
    --eval.interval 5 \
        --eval.initial_validation False \
        --eval.use_sample_metric False \
    --kv_cache.cache_length 32768 \
        --kv_cache.chunk_size 2048 \
        --kv_cache.name lastrec-torch-quantized8 \
    --grad.layers_per_cell 2 \
        --grad.layercp_qname default \
        --grad.cachecp_qname torch-quantized8 \
    --optimizer.name AdamW \
        --optimizer.learning_rate 0.0005 \
    --resume step-000015

[TODO]
```

**Comment by mseeger** [comment]
We see that data batches are different for iterations when training is resumed.

But this is not due to the training state not being stored/loaded properly. But instead, the dataset cases are created differently by the data loader. This is odd, because dataset intermediates seem to be cached and loaded from cache on disk.
